gpflow/models/training_interface.py

```diff
@@ -0,0 +1,62 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData)
+
+
+class TrainingInterface:
+    @abc.abstractmethod
+    def training_loss(self, *args, **kwargs) -> tf.Tensor:
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def training_loss_closure(self, *args, **kwargs):
+        """
+        Return a closure with a training loss for this module
+        """
+        raise NotImplementedError
+
+
+class InternalDataTrainingInterface(TrainingInterface):
+    def training_loss(self):
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingInterface(TrainingInterface):
+    def training_loss(self, data):
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    def training_loss_closure(self, data: Data, jit=True) -> Callable[[], tf.Tensor]:
+        training_loss = self.training_loss
+        if jit:
+            training_loss = tf.function(
+                self.training_loss
+            )  # TODO need to add correct input_signature here to allow for differently sized minibatches
```

Comment:
how would you do this? By having a user specify `function_kwargs` as argument?

Reply:
i think you just specifiy the leading dim size to be None

Reply:
specify `input_signatures` argument to the tf.function - just have to look up the right TensorSpec ...

Reply:
this won't work in a general way for models that have different shapes of data -- e.g. RegressionData (SVGP) with data=(X,Y) vs InputData with data=Y (if there was a minibatching GPLVM) need different (incompatible) input_signatures.